xdsl/dialects/smt.py

```diff
@@ -117,6 +129,29 @@ def __init__(
         )
 
 
+@irdl_op_definition
+class ApplyFuncOp(IRDLOperation):
+    """
+    This operation performs a function application as described in the SMT-LIB
+    2.7 standard. It is part of the SMT-LIB core theory.
+    """
+
+    name = "smt.apply_func"
+
+    DOMAIN: ClassVar = RangeVarConstraint("DOMAIN", RangeOf(base(NonFuncSMTType)))
+    RANGE: ClassVar = VarConstraint("RANGE", base(NonFuncSMTType))
+
+    func = operand_def(FuncType.constr(DOMAIN, RANGE))
+    args = var_operand_def(DOMAIN)
+
+    result = result_def(RANGE)
+
+    assembly_format = "$func `(` $args `)` attr-dict `:` type($func)"
+
+    def __init__(self, func: SSAValue[FuncType], args: Sequence[SSAValue]):
```

Comment:
Let me know what you think of this.
I require a `SSAValue[FuncType]` as input, as I need to grab its range type.
I feel that now that we have generic types, this interface makes more sense than requiring an `SSAValue`, and then having an `if isinstance(value.type, FuncType)` with a custom error message.
What do you think @alexarice @superlopuh ?

Reply:
Looks good, this was the point of the typed SSAValues right?

Reply:
Oh yeah my bad, I forgot about that one!

Reply:
Yeah I like it

Reply:
Done!